File: CSM_capcal__01_cin_06.m
%% Sol·lució implementada trajectòria capçal
% Generació de trajectòries
% 01_cin_06.tex
% J.A.Mayugo, UdG, 2020

close all; clear;
%% DADES
t_total = 8.5;  % segons
a_max = 200;    % mm/s2

rP = 80/2;      % mm
i = 66;

%% Resolució: Métode 1, marxa-moviment-parada
% S'imposa un temps de treball i una velocitat de treball a l'anada
t_treball = 4;      % segons
v_treball = 100;    % mm/s

% Paràmetres cicle d'anada
tb_a    = v_treball/a_max;
xb_a    = 1/2*a_max*tb_a^2;

ta      = 2*tb_a + t_treball;
xa      = 2*xb_a + t_treball*v_treball;

% Paràmetres cicle de tornada
tt      = t_total - ta;
tb_t    = tt/2 - sqrt(a_max^2*tt^2-4*a_max*(xa))/2/a_max;
xb_t    = 1/2*a_max*tb_t^2;

v_max   = tb_t*a_max;

% En el motor:
omega_max       = v_max*i/rP
omega_treball   = v_treball*i/rP

n_max           = omega_max*60/(2*pi)
n_treball       = omega_treball*60/(2*pi)

% Definir trams de moviment: motion control
ti(1) = tb_a;
ti(2) = ti(1) + t_treball;
ti(3) = ti(2) + tb_a;
ti(4) = ti(3) + tb_t;
ti(5) = ti(4) + tt-2*tb_t;
ti(6) = ti(5) + tb_t;

xi(1) =       + 1/2*a_max*ti(1)^2;
xi(2) = xi(1) + v_treball*(ti(2)-ti(1));
xi(3) = xi(2) + v_treball*(ti(3)-ti(2)) - 1/2*a_max*(ti(3)-ti(2))^2;
xi(4) = xi(3) + 0*(ti(4)-ti(3)) - 1/2*a_max*(ti(4)-ti(3))^2;
xi(5) = xi(4) - v_max*(ti(5)-ti(4));
xi(6) = xi(5) - v_max*(ti(6)-ti(5)) + 1/2*a_max*(ti(6)-ti(5))^2;

% Obtenir la trajectoria posicio, velocitat i acceleració
t_ = [0:0.05:t_total];
x_ = zeros(1,length(t_));
v_ = zeros(1,length(t_));
a_ = zeros(1,length(t_));

for ii = [1:length(t_)]
    if  t_(ii) <= ti(1)
        x_(ii) = + 1/2*a_max*t_(ii)^2;
        v_(ii) = a_max*t_(ii);
        a_(ii) = a_max;
    elseif  t_(ii) <= ti(2)
        x_(ii) = xi(1) + v_treball*(t_(ii)-ti(1));
        v_(ii) = v_treball;
        a_(ii) = 0;
    elseif  t_(ii) <= ti(3)
        x_(ii) = xi(2) + v_treball*(t_(ii)-ti(2)) - 1/2*a_max*(t_(ii)-ti(2))^2;
        v_(ii) = v_treball - a_max*(t_(ii)-ti(2));
        a_(ii) = -a_max;
    elseif  t_(ii) <= ti(4)
        x_(ii) = xi(3) + 0*(t_(ii)-ti(3)) - 1/2*a_max*(t_(ii)-ti(3))^2;
        v_(ii) = 0 - a_max*(t_(ii)-ti(3));
        a_(ii) = -a_max;
    elseif  t_(ii) <= ti(5)
        x_(ii) = xi(4) - v_max*(t_(ii)-ti(4));
        v_(ii) = -v_max;
        a_(ii) = 0;
    else
        x_(ii) = xi(5) - v_max*(t_(ii)-ti(5)) + 1/2*a_max*(t_(ii)-ti(5))^2;
        v_(ii) = -v_max + a_max*(t_(ii)-ti(5));
        a_(ii) = a_max;
    end
end

% Representar poció, velocitat i acceleració
H1=figure;font_size=18;
subplot(3,1,1);
plot(t_,x_,'LineWidth',2)
ylabel('${u}$ [mm]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
subplot(3,1,2);
plot(t_,v_,'LineWidth',2)
ylabel('$\dot{u}$ [mm/s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
subplot(3,1,3);
plot(t_,a_,'LineWidth',2)
ylabel('$\ddot{u} \; \rm{[mm/s^2]}$','Interpreter','latex','FontSize',font_size)
xlabel('$t$ [s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
%saveas(H1,'./figures/CIN_Ex11_met1.svg')
%saveas(H1,'./figures/CIN_Ex11_met1.pdf')

% Representar omega motor
H11=figure;font_size=18;
plot(t_,v_*i/rP,'LineWidth',2)
ylabel('$\omega_m$ [rad/s]','Interpreter','latex','FontSize',font_size)
xlabel('$t$ [s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
H11.Position = [100 100 560 240];
%saveas(H11,'./figures/CIN_Ex11_met1_omg.svg')
%saveas(H11,'./figures/CIN_Ex11_met1_omg.pdf')

%% Resolució: Métode 2, definir trajectòria punt a punt
% S'imposa arranc inicial, llavors un moviment de treball i un de tornada
% definint 4 punts

td12 = 0.4;  % segons
td23 = 4.9;  % segons
td34 = t_total-(td12+td23); 

x1 = 0;      % mm
x2 = 10;     % mm
x3 = 500;    % mm
x4 = 0;      % mm

% Paràmetres trams
% tram de 1 a 2 (inicial)
a1 =  (x2-x1)/abs(x2-x1)*a_max;
t1 = td12 - (td12^2-2*(x2-x1)/a1)^0.5;
v12 = (x2-x1)/(td12-1/2*t1);

% tram de 3 a 4 (final)
a4 = (x3-x4)/abs(x4-x3)*a_max;
t4 = td34 - (td34^2+2*(x4-x3)/a4)^0.5;
v34 = (x4-x3)/(td34-1/2*t4);

% VELOCITATS
% tram de 2 a 3
v23 = (x3-x2)/td23;

% ACCELERACIONS
% tram 2
a2 = (v23-v12)/abs(v12-v23)*a_max;
% tram 3
a3 = (v34-v23)/abs(v23-v34)*a_max;

% intervals de temps
t2 = (v23-v12)/a2;
t3 = (v34-v23)/a3;
t12 = td12 - t1 - 1/2*t2;
t23 = td23 - 1/2*(t2+t3);
t34 = td34  - t4 - 1/2*t3;
ttotal2=t1+t2+t3+t4+t12+t23+t34

% intervals de POSICIÓ
xi1 = 0 *t1  + 1/2*a1*t1^2;
xi2 = v12*t2 + 1/2*a2*t2^2;
xi3 = v23*t3 + 1/2*a3*t3^2;
xi4 = v34*t4 + 1/2*a4*t4^2;
x12 = v12*t12;
x23 = v23*t23;
x34 = v34*t34;
xtotal2 = xi1 + xi2+xi3+xi4+x12+x23+x34

% 		Tram &  Tipus  & Duració  &  Despl.  &Velocitat & Acceleració    \\
% 		$(i)$ &   & $t$[ s] & ${x}$ [mm]& $\dot{x}$ [mm/s] & $\ddot{x}$ [mm/s\tss{2}]    \\
disp(['Tram 1 , MUA',', t_1 =',num2str(t1,'%10.3f'),'s , x_1 =',num2str(xi1,'%10.3f') 'mm,          a_1 =',num2str(a1,'%10.f'),'mm/s2' ])
disp(['Tram 12, MU ',', t_12=',num2str(t12,'%10.3f'),'s , x_12=',num2str(x12,'%10.3f') 'mm, v_12=',num2str(v12,'%10.f'),'mm/s' ])
disp(['Tram 2 , MUA',', t_2 =',num2str(t2,'%10.3f'),'s , x_2 =',num2str(xi2,'%10.3f') 'mm,          a_2 =',num2str(a2,'%10.f'),'mm/s2' ])
disp(['Tram 23, MU ',', t_23=',num2str(t23,'%10.3f'),'s , x_23=',num2str(x23,'%10.3f') 'mm, v_23=',num2str(v23,'%10.f'),'mm/s' ])
disp(['Tram 3 , MUA',', t_3 =',num2str(t3,'%10.3f'),'s , x_3 =',num2str(xi3,'%10.3f') 'mm,          a_3 =',num2str(a3,'%10.f'),'mm/s2' ])
disp(['Tram 34, MU ',', t_34=',num2str(t34,'%10.3f'),'s , x_34=',num2str(x34,'%10.3f') 'mm, v_34=',num2str(v34,'%10.f'),'mm/s' ])
disp(['Tram 4 , MUA',', t_4 =',num2str(t4,'%10.3f'),'s , x_4 =',num2str(xi4,'%10.3f') 'mm,          a_4 =',num2str(a4,'%10.f'),'mm/s2' ])

disp(' ')
disp(['temps total = =',num2str(ttotal2,'%10.3f'),'s , x total =',num2str(abs(xtotal2),'%10.3f'), 'mm'])

% Definir trams de moviment: motion control
ti(1)=      t1;
ti(2)=ti(1)+t12;
ti(3)=ti(2)+t2;
ti(4)=ti(3)+t23;
ti(5)=ti(4)+t3;
ti(6)=ti(5)+t34;
ti(7)=ti(6)+t4;

xi(1)=      xi1;
xi(2)=xi(1)+x12;
xi(3)=xi(2)+xi2;
xi(4)=xi(3)+x23;
xi(5)=xi(4)+xi3;
xi(6)=xi(5)+x34;
xi(7)=xi(6)+xi4;

% Obtenir la trajectoria posicio, velocitat i acceleració
t_ = [0:0.05:t_total];
x_ = zeros(1,length(t_));
v_ = zeros(1,length(t_));
a_ = zeros(1,length(t_));

for ii = [1:length(t_)]
    if  t_(ii) <= ti(1)
        x_(ii) = 0 + 1/2*a1*t_(ii)^2;
        v_(ii) = a1*t_(ii);
        a_(ii) = a1;
    elseif  t_(ii) <= ti(2)
        x_(ii) = xi(1) + v12*(t_(ii)-ti(1));
        v_(ii) = v12;
        a_(ii) = 0;
    elseif    t_(ii) <= ti(3)
        x_(ii) = xi(2) + v12*(t_(ii)-ti(2)) + 1/2*a2*(t_(ii)-ti(2))^2;
        v_(ii) = v12 + a2*(t_(ii)-ti(2));
        a_(ii) = a2;
    elseif  t_(ii) <= ti(4)
        x_(ii) = xi(3) + v23*(t_(ii)-ti(3));
        v_(ii) = v23;
        a_(ii) = 0;
    elseif    t_(ii) <= ti(5)
        x_(ii) = xi(4) + v23*(t_(ii)-ti(4)) + 1/2*a3*(t_(ii)-ti(4))^2;
        v_(ii) = v23 + a3*(t_(ii)-ti(4));
        a_(ii) = a3;
    elseif  t_(ii) <= ti(6)
        x_(ii) = xi(5) + v34*(t_(ii)-ti(5));
        v_(ii) = v34;
        a_(ii) = 0;
    else
        x_(ii) = xi(6) + v34*(t_(ii)-ti(6)) + 1/2*a4*(t_(ii)-ti(6))^2;
        v_(ii) = v34 + a4*(t_(ii)-ti(6));
        a_(ii) = a4;
    end
end

% Representar poció, velocitat i acceleració
H2=figure;
subplot(3,1,1);
plot(t_,x_,'LineWidth',2)
ylabel('$u$ [mm]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
subplot(3,1,2);
plot(t_,v_,'LineWidth',2)
ylabel('$\dot{u}$ [mm/s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
subplot(3,1,3);
plot(t_,a_,'LineWidth',2)
ylabel('$\ddot{u} \; \rm{[mm/s^2]}$','Interpreter','latex','FontSize',font_size)
xlabel('$t$ [s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
%saveas(H2,'./figures/CIN_Ex11_met2.svg')
%saveas(H2,'./figures/CIN_Ex11_met2.pdf')

% Representar omega motor
H21=figure;font_size=18;
plot(t_,v_*i/rP,'LineWidth',2)
ylabel('$\omega_m$ [rad/s]','Interpreter','latex','FontSize',font_size)
xlabel('$t$ [s]','Interpreter','latex','FontSize',font_size)
set(gca,'FontSize',font_size)
xlim([0 t_total])
H21.Position = [100 100 560 240];
%saveas(H21,'./figures/CIN_Ex11_met2_omg.svg')
%saveas(H21,'./figures/CIN_Ex11_met2_omg.pdf')

File: CSM_capcal__01_cin_06.py
import numpy as np
import matplotlib.pyplot as plt

# %%
# ======================
# DADES
# ======================
t_total = 8.5      # s, temps total de cicle
a_max = 200        # mm/s^2, acceleració màxima

rP = 80 / 2        # mm, radi politja
i = 66

# %%
# ======================
# Resolució: Métode 1, marxa-moviment-parada
# ======================
# S'imposa un temps de treball i una velocitat de treball a l'anada
t_treball = 4       # s, temps de treball
v_treball = 100     # mm/s, velocitat de treball

# Paràmetres del cicle d'anada (ta)
tb_a = v_treball / a_max
xb_a = 0.5 * a_max * tb_a**2

ta = 2 * tb_a + t_treball
xa = 2 * xb_a + t_treball * v_treball

# Paràmetres del cicle de tornada (tt)
tt = t_total - ta

tb_t = (tt/2) - np.sqrt(a_max**2 * tt**2 - 4 * a_max * xa) / (2 * a_max)
xb_t = 0.5 * a_max * tb_t**2

v_max = tb_t * a_max

# En el motor
omega_max = v_max * i / rP
omega_treball = v_treball * i / rP

n_max = omega_max * 60 / (2*np.pi)
n_treball = omega_treball * 60 / (2*np.pi)

# % Definir trams de moviment: motion control
ti = np.zeros(6)
ti[0] = tb_a
ti[1] = ti[0] + t_treball
ti[2] = ti[1] + tb_a
ti[3] = ti[2] + tb_t
ti[4] = ti[3] + tt - 2*tb_t
ti[5] = ti[4] + tb_t

xi = np.zeros(6)
xi[0] = 0.5 * a_max * ti[0]**2
xi[1] = xi[0] + v_treball * (ti[1] - ti[0])
xi[2] = xi[1] + v_treball * (ti[2] - ti[1]) - 0.5 * a_max * (ti[2] - ti[1])**2
xi[3] = xi[2] - 0.5 * a_max * (ti[3] - ti[2])**2
xi[4] = xi[3] - v_max * (ti[4] - ti[3])
xi[5] = xi[4] - v_max * (ti[5] - ti[4]) + 0.5 * a_max * (ti[5] - ti[4])**2

# % Obtenir la trajectoria posicio, velocitat i acceleració
t_ = np.arange(0, t_total, 0.05)
x_ = np.zeros_like(t_)
v_ = np.zeros_like(t_)
a_ = np.zeros_like(t_)

for k, t in enumerate(t_):
    if t <= ti[0]:
        x_[k] = 0.5 * a_max * t**2
        v_[k] = a_max * t
        a_[k] = a_max

    elif t <= ti[1]:
        x_[k] = xi[0] + v_treball * (t - ti[0])
        v_[k] = v_treball

    elif t <= ti[2]:
        x_[k] = xi[1] + v_treball*(t - ti[1]) - 0.5*a_max*(t - ti[1])**2
        v_[k] = v_treball - a_max*(t - ti[1])
        a_[k] = -a_max

    elif t <= ti[3]:
        x_[k] = xi[2] - 0.5*a_max*(t - ti[2])**2
        v_[k] = -a_max*(t - ti[2])
        a_[k] = -a_max

    elif t <= ti[4]:
        x_[k] = xi[3] - v_max*(t - ti[3])
        v_[k] = -v_max

    else:
        x_[k] = xi[4] - v_max*(t - ti[4]) + 0.5*a_max*(t - ti[4])**2
        v_[k] = -v_max + a_max*(t - ti[4])
        a_[k] = a_max

# %%
#  Representar poció, velocitat i acceleració
plt.figure(figsize=(8,6))

plt.subplot(3,1,1)
plt.plot(t_, x_)
plt.ylabel(r"$u$ [mm]")

plt.subplot(3,1,2)
plt.plot(t_, v_)
plt.ylabel(r"$v$ [mm/s]")

plt.subplot(3,1,3)
plt.plot(t_, a_)
plt.ylabel(r"$a$ [mm/s²]")
plt.xlabel(r"$t$ [s]")

plt.tight_layout()
plt.show()

plt.figure(figsize=(8,3))
plt.plot(t_, v_ * i / rP)
plt.ylabel(r"$\omega_m$ [rad/s]")
plt.xlabel("$t$ [s]")
plt.show()

# %%
# =========================================================
# Resolució: Métode 2, definir trajectòria punt a punt
# =========================================================
# S'imposa arranc inicial, llavors un moviment de treball i un de tornada
# definint 4 punts
td12 = 0.4
td23 = 4.9
td34 = t_total - (td12 + td23)

x1, x2, x3, x4 = 0, 10, 500, 0    # mm, 

## Paràmetres trams
# % tram de 1 a 2 (inicial)
a1 = np.sign(x2-x1) * a_max
t1 = td12 - np.sqrt(td12**2 - 2*(x2-x1)/a1)
v12 = (x2-x1)/(td12 - 0.5*t1)

# % tram de 3 a 4 (final)
a4 = np.sign(x3-x4) * a_max
t4 = td34 - np.sqrt(td34**2 + 2*(x4-x3)/a4)
v34 = (x4-x3)/(td34 - 0.5*t4)

# % VELOCITATS: tram de 2 a 3
v23 = (x3-x2)/td23

# % ACCELERACIONS: tram 2
a2 = np.sign(v23-v12) * a_max
a3 = np.sign(v34-v23) * a_max

# % intervals de temps
t2 = (v23 - v12)/a2
t3 = (v34 - v23)/a3
t12 = td12 - t1 - 0.5*t2
t23 = td23 - 0.5*(t2 + t3)
t34 = td34 - t4 - 0.5*t3
ttotal2=t1+t2+t3+t4+t12+t23+t34

# % intervals de POSICIÓ
xi1 = 0 *t1  + 1/2*a1*t1**2;
xi2 = v12*t2 + 1/2*a2*t2**2;
xi3 = v23*t3 + 1/2*a3*t3**2;
xi4 = v34*t4 + 1/2*a4*t4**2;
x12 = v12*t12;
x23 = v23*t23;
x34 = v34*t34;
xtotal2 = xi1 + xi2+xi3+xi4+x12+x23+x34

print(f"Tram 1 , MUA, t_1 = {t1:10.3f} s , x_1 = {xi1:10.3f} mm, a_1 = {a1:10.0f} mm/s^2")
print(f"Tram 12, MU, t_12 = {t12:10.3f} s , x_12 = {x12:10.3f} mm, v_12 = {v12:10.0f} mm/s")
print(f"Tram 2 , MUA, t_2 = {t2:10.3f} s , x_2 = {xi2:10.3f} mm, a_2 = {a2:10.0f} mm/s^2")
print(f"Tram 23, MU, t_23 = {t23:10.3f} s , x_23 = {x23:10.3f} mm, v_23 = {v23:10.0f} mm/s")
print(f"Tram 3 , MUA, t_3 = {t3:10.3f} s , x_3 = {xi3:10.3f} mm, a_3 = {a3:10.0f} mm/s^2")
print(f"Tram 34, MU, t_34 = {t34:10.3f} s , x_34 = {x34:10.3f} mm, v_34 = {v34:10.0f} mm/s")
print(f"Tram 4 , MUA, t_4 = {t4:10.3f} s , x_4 = {xi4:10.3f} mm, a_4 = {a4:10.0f} mm/s^2")
print(" ")
print(f"Temps total = {ttotal2:10.3f} s , x total = {abs(xtotal2):10.3f} mm")

# %%
ti2 = np.zeros(7)
ti2[0] = t1
ti2[1] = ti2[0] + t12
ti2[2] = ti2[1] + t2
ti2[3] = ti2[2] + t23
ti2[4] = ti2[3] + t3
ti2[5] = ti2[4] + t34
ti2[6] = ti2[5] + t4

xi_2 = np.zeros(7)
xi_2[0] = xi1
xi_2[1] = xi_2[0] + x12
xi_2[2] = xi_2[1] + xi2
xi_2[3] = xi_2[2] + x23
xi_2[4] = xi_2[3] + xi3
xi_2[5] = xi_2[4] + x34
xi_2[6] = xi_2[5] + xi4

# % Obtenir la trajectoria posicio, velocitat i acceleració
t_ = np.arange(0, t_total, 0.05)
x2_ = np.zeros_like(t_)
v2_ = np.zeros_like(t_)
a2_ = np.zeros_like(t_)

for k, t in enumerate(t_):
    if t <= ti2[0]:
        x2_[k] = 0 + 0.5*a1*t**2
        v2_[k] = a1*t
        a2_[k] = a1
    elif t <= ti2[1]:
        x2_[k] = xi_2[0] + v12*(t - ti2[0])
        v2_[k] = v12
        a2_[k] = 0
    elif t <= ti2[2]:
        x2_[k] = xi_2[1] + v12*(t - ti2[1]) + 0.5*a2*(t - ti2[1])**2
        v2_[k] = v12 + a2*(t - ti2[1])
        a2_[k] = a2
    elif t <= ti2[3]:
        x2_[k] = xi_2[2] + v23*(t - ti2[2])
        v2_[k] = v23
        a2_[k] = 0
    elif t <= ti2[4]:
        x2_[k] = xi_2[3] + v23*(t - ti2[3]) + 0.5*a3*(t - ti2[3])**2
        v2_[k] = v23 + a3*(t - ti2[3])
        a2_[k] = a3
    elif t <= ti2[5]:
        x2_[k] = xi_2[4] + v34*(t - ti2[4])
        v2_[k] = v34
        a2_[k] = 0
    else:
        x2_[k] = xi_2[5] + v34*(t - ti2[5]) + 0.5*a4*(t - ti2[5])**2
        v2_[k] = v34 + a4*(t - ti2[5])
        a2_[k] = a4

# %%
# % Representar poció, velocitat i acceleració
plt.figure(figsize=(8,6))
plt.subplot(3,1,1)
plt.plot(t_, x2_)
plt.ylabel(r"$u$ [mm]")

plt.subplot(3,1,2)
plt.plot(t_, v2_)
plt.ylabel(r"$v$ [mm/s]")

plt.subplot(3,1,3)
plt.plot(t_, a2_)
plt.ylabel(r"$a$ [mm/s²]")
plt.xlabel(r"$t$ [s]")

plt.tight_layout()
plt.show()